airflow/dag_processing/processor.py

```diff
@@ -149,8 +149,8 @@ def _run_file_processor(
 
         try:
             # redirect stdout/stderr to log
-            with redirect_stdout(StreamLogWriter(log, logging.INFO)), redirect_stderr(
-                StreamLogWriter(log, logging.WARN)
+            with redirect_stdout(StreamLogWriter(log, logging.INFO)), (  # type: ignore[type-var]
+                redirect_stderr(StreamLogWriter(log, logging.WARN))  # type: ignore[type-var]
```

Comment:
Errors here:
```
airflow/dag_processing/processor.py:152: error: Value of type variable "_T_io"
of "redirect_stdout" cannot be "StreamLogWriter"
                with redirect_stdout(StreamLogWriter(log, logging.INFO)), ...
                     ^
airflow/dag_processing/processor.py:152: error: Value of type variable "_T_io"
of "redirect_stderr" cannot be "StreamLogWriter"
```

Typing for args of both `redirect_stdout()` and `redirect_stderr()` is handled in an external
contextlib stub file and defined as `ContextManager[TypeVar("_T_io", bound=Optional[IO[str]])]`
so ignoring the Mypy errors here. But any feedback on a proper handling of this is welcomed of course.

Reply:
We can implement the `IO` protocol on `StreamLogWriter` instead. It already has most of the important things implemented, the rest are mostly trivial.

Reply:
I pushed a fix (which is really a workaround). I could not find to make it works but It seems that unless we want to reimplement all the methods manually, the IOBase and IO type hierarchies are incompatible and other than ignoring the MyPy errors there is not much we can do about it 

https://giters.com/python/typeshed/issues/6077

Reply:
At least I pushed it down as much as I could. The ignores now simply ignore those errors:

```
airflow/utils/log/logging_mixin.py:83: error: Definition of "__next__" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "__next__" in base class "IOBase" is incompatible with definition in base class "Iterator"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "readlines" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "__iter__" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "__iter__" in base class "IOBase" is incompatible with definition in base class "Iterator"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "__iter__" in base class "IOBase" is incompatible with definition in base class "Iterable"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "readline" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "writelines" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^
airflow/utils/log/logging_mixin.py:83: error: Definition of "__enter__" in base class "IOBase" is incompatible with definition in base class "IO"
    class StreamLogWriter(IOBase, IO[str]):
    ^

```

Reply:
Or maybe there is an IO Protocol you mentioned @uranusjr which I simply could not find ?  The only "IO" I see is the Generic (which I now implemented) - but being Generic it does not have any implementation. The only "base" implementation I am aware of that implements IO is  BaseIO family (but then it's incompatible with the Generic).

Reply:
I am also not 100% sure if removing encoding = None is a good thing and whether it does not have any side effects :(.